starlite/middleware/session.py

```diff
@@ -0,0 +1,233 @@
+import binascii
+import contextlib
+import time
+from base64 import b64decode, b64encode
+from os import urandom
+from typing import TYPE_CHECKING, Any, Awaitable, Callable, List, Optional
+
+from orjson import OPT_SERIALIZE_NUMPY, dumps
+from orjson.orjson import loads
+from pydantic import (
+    BaseConfig,
+    BaseModel,
+    SecretBytes,
+    conint,
+    conlist,
+    constr,
+    validator,
+)
+from starlette.datastructures import MutableHeaders
+from starlette.requests import HTTPConnection
+from typing_extensions import Literal
+
+from starlite.datastructures import Cookie
+from starlite.exceptions import MissingDependencyException
+from starlite.middleware.base import MiddlewareProtocol
+from starlite.response import Response
+
+try:
+    from cryptography.exceptions import InvalidTag
+    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+except ImportError as e:
+    raise MissingDependencyException("cryptography is not installed") from e
+
+if TYPE_CHECKING:
+    from starlette.types import ASGIApp, Message, Receive, Scope, Send
+
+ONE_DAY_IN_SECONDS = 60 * 60 * 24
+NONCE_SIZE = 12
+CHUNK_SIZE = 4096 - 512
+AAD = b"additional_authenticated_data="
+
+
+class SessionCookieConfig(BaseModel):
+    """Configuration for Session middleware."""
+
+    class Config(BaseConfig):
+        arbitrary_types_allowed = True
+
+    secret: SecretBytes
+    """
+    A secret key to use for generating an encryption key.
+    Must have a length of 16 (128 bits), 24 (192 bits) or 32 (256 bits) characters.
+    """
+    key: constr(min_length=1, max_length=256) = "session"  # type: ignore[valid-type]
+    """
+    Key to use for the cookie inside the header,
+    e.g. `session=<data>` where 'session' is the cookie key and <data> is the session data.
+
+    Notes:
+        - If a session cookie exceeds 4KB in size it is split. In this case the key will be of the format
+            'session-{segment number}'.
+    """
+    max_age: conint(ge=1) = ONE_DAY_IN_SECONDS * 14  # type: ignore[valid-type]
+    """Maximal age of the cookie before its invalidated."""
+    scopes: conlist(Literal["http", "websocket"], min_items=1, max_items=2) = ["http", "websocket"]  # type: ignore[valid-type]
+    """Scopes for the middleware - options are 'http' and 'websocket' with the default being both"""
+    path: str = "/"
+    """Path fragment that must exist in the request url for the cookie to be valid. Defaults to '/'."""
+    domain: Optional[str] = None
+    """Domain for which the cookie is valid."""
+    secure: Optional[bool] = None
+    """Https is required for the cookie."""
+    httponly: Optional[bool] = True
```

Comment:
I don't think this needs to be `Optional` as explicitly setting this value to `None` has the same outcome as setting it to `False` right? Just keeps the api that little bit simpler IMO.